Categorize match where

hvub opened the pull request:
This PR categorized scenarios testing the `WHERE` subclause of the `MATCH` clause.

One scenario has been added to [Match7.[25]](https://github.com/opencypher/openCypher/compare/master...hvub:categorize-match-where?expand=1#diff-c798c3cedb1f755db1567273da1a91bfR536) and three subsequent scenario add been renumbered.

27 scenarios have been added to `match-where`. Most of these have been renamed fro more clarity.

In total, 5 scenarios have been added, based in variations of existing scenarios.

DvirDukhan commented on tck/features/precategorized/clauses/where/Match7-2.feature:
```diff
@@ -30,13 +30,15 @@
 
-  Scenario: MATCH and OPTIONAL MATCH on same pattern
+  # Consider for OptionalMatch-Where1
+  Scenario: Filter node with node label predicate on multi-column binding table after MATCH and OPTIONAL MATCH
```
binding table is internal

DvirDukhan commented on tck/features/precategorized/clauses/where/Match7-2.feature:
```diff
@@ -65,15 +68,17 @@ Feature: Match7-2 - Optional match WHERE clause scenarios
 
-  Scenario: Return null when no matches due to label predicate in WHERE
+  # Consider for OptionalMatch-Where1
+  Scenario: Filter node with property predicate on multi-column binding table after OPTIONAL MATCH
```
binding table is internal

hvub commented on tck/features/clauses/match-where/MatchWhere1.feature:
```diff
@@ -28,26 +28,59 @@
+Feature: MatchWhere1 - Filter single variable
 
-  Scenario: Filter out based on node prop name
+  Scenario: [1] Filter node with node label predicate on multi-column binding table
```
Conceptually, `MATCH` produces a binding table and `WHERE` filters this binding table. The binding table has as many columns as where are distinct variables. In this scenario, `MATCH` produces a two-column binding table (columns `a` and `b`) and `WHERE` filters this binding table with a node label predicate (`a:A`). Makes sense?

hvub replied:
No, binding table the concept that we use to describe the semantics of the language. The GQL specification also uses this concept. In fact, an implementation likely rolls the evaluation an certain `WHERE` predicates directly into pattern matching.

hvub commented on tck/features/clauses/match-where/MatchWhere1.feature:
```diff
@@ -28,26 +28,59 @@
+      | 0    | 1    |
+    And no side effects
+
+  Scenario: [2] Filter node with node label predicate on empty multi-column binding table
```
Same as above, except that the binding table is empty since the `MATCH` will not match anything on the given graph. The scenarios may look odd, but it is a useful test case to check that `WHERE` is not troubles by an empty binding table and does not add any rows of columns to it, I guess.

hvub commented on tck/features/clauses/match-where/MatchWhere1.feature:
```diff
@@ -28,26 +28,59 @@
     And no side effects
 
-  Scenario: Matching using a simple pattern with label predicate
+  Scenario: [3] Filter node with property predicate on single-column binding table
```
Let's use the term "single/mutliple variable bindings" instead.

hvub commented:
@DvirDukhan I have renamed the said scenarios towards using "on single/multi variables with one binding/with multiple bindings/without any bindings" to reflect number of column and number of rows the `WHERE` has to deal with.

I also stumbled over one scenarios wrongly categorized scenario and a pre-categorized sibling scenario and moved both to a new pre-categorized feature, cf. `tck/features/precategorized/clauses/with-where/WithWhere1.feature`

DvirDukhan commented:
@hvub you can merge this. Looks great.
